docs/config.md
# Configuration

Created via `./bin/console config:dump-reference DbpRelayCoreBundle | sed '/^$/d'`

```yaml
# Default configuration for "DbpRelayCoreBundle"
dbp_relay_core:
    # Some string identifying the current build (commit hash)
    build_info:           ~ # Example: deadbeef
    # Some URL identifying the current build (URL to the commit on some git web interface)
    build_info_url:       ~ # Example: 'https://github.example.com/project/-/commit/deadbeef'
    # Path to the logo (256x256) of the API frontend
    logo_path:            ~ # Example: bundles/dbprelaycore/logo.png
    # The title text of the API docs page
    docs_title:           'Relay API Gateway'
    # The description text of the API docs page (supports markdown)
    docs_description:     '*part of the [digital blueprint](https://www.digital-blueprint.org) project*'
    # See https://symfony.com/doc/5.3/messenger.html#redis-transport
    queue_dsn:            '' # Example: 'redis://redis:6379'
    # https://symfony.com/doc/5.3/components/lock.html
    lock_dsn:             '' # Example: 'redis://redis:6379'
```


CHANGELOG.md
# Changelog

## Unreleased

## v0.1.227

- TestUtils: allow passing the URI variables of the current request to
DataProviderTester and DataProcessorTester

## v0.1.226

- Add support for Symfony 7.4
- Add support for symfony/monolog-bundle 4
- Remove various deprecated classes and methods:
  - AbstractEntityDeNormalizer
  - AbstractGetAttributeSubscriber
  - Tools::decodeJSON
  - Tools::dumpTrace
  - AbstractStateProvider
  - NamedEntityInterface

## v0.1.225

- don't ship all subclasses of `AbstractAuthorizationService` with an `EntityNormalizer`
- implement `ResetInterface::reset` where possible for kernel testing
- offer `TestAuthorizationService::reset` for manual reset of request caches for non-kernel tests
- remove deprecate `policies` from authorization config in favor of `roles` and `resource_permissions` (deprecated since v0.1.188)

## v0.1.224

- deprecate `UserAuthTrait` in favor of `TestClient`

## v0.1.223

- Inject the `EntityNormalizer` into `AbstractAuthorizationService` for testing to avoid null references
- TestClient: offer a method to reset the request caches of a given `AbstractAuthorizationService` service

## v0.1.222

- Restore missing dependency on symfony/asset

## v0.1.221

- api-docs: update vendored bundle, move from rollup to rolldown for bundling
- Drop support for api-platform v3
- Move dependencies to split api-platform packages

## v0.1.220

- Add support for kevinrob/guzzle-cache-middleware v7, drop support for v4

## v0.1.219

- Allow configuration of identifier name for `DataProviderTester` and `DataProcessorTester`
- Remove strict dependency on doctrine migrations bundle

## v0.1.218

- Allow HEAD for CORS requests as well
- Drop workaround for broken CORS headers with Apache <=2.4.47 for 304 responses
  since no one should be using such old Apache versions anymore.

## v0.1.217

- AbstractAuthorizationService: add authorization expressions on setupAccessControlPolicies instead of replacing them 
(allows subclasses of AbstractDataProvider to define authorization expressions without being replaced internally)
- AbstractLocalDataEventSubscriber: map local data attribute paths in filter to their respective source attribute paths
- add FilterTools
- Decorate the 'doctrine.migrations.migrations_factory' service to be able to inject the service container into doctrine
migration subclasses. 
- Add AbstractEntityManagerMigration which gets injected the service container and is meant to be subclassed by doctrine migrations versions

## v0.1.215

- Rework user attribute provision: Change UserAttributeProviderInterface from "get-all-available-attributes" to "has-get-attribute"

## v0.1.214

- Deprecate QueryHelper::getEntities and add  QueryHelper::getEntitiesPageNumberBased and QueryHelper::getEntitiesPageStartIndexBased and 

## v0.1.213

- rest/query/filters: validate filter operators and values on construction of ConditionNode

## v0.1.212

- deprecate Pagination::getAllResults and introduce Pagination::getAllResultsPageNumberBased and Pagination::getAllResultsPageStartIndexBased

## v0.1.211

- Fix an internal error when requesting /errors/XXX

## v0.1.210

- support both api-platform 3 and 4

## v0.1.209

- add MigratePostEvent that is fired after a successful DB migration
- support both api-platform 3 and 4

## v0.1.208

- Add support for Api-Platform v4

## v0.1.207

- filter query parameter syntax: require string values to be quoted with double quotes to be able to
distinguish between strings and numeric/boolean values and be type safe
- force use filters (where `force_use_policy` evaluates to `true`) for get item operations as well,
for get item and get collection to be consistent. NOTE: implementors are responsible to apply those filters
in their overrides of `getItemById` (filters are found in the `$options` parameter, like for `getPage`).

## v0.1.206

- Drop support for PHP 8.1
- Drop support for Symfony 5
- Drop support for api-platform 3.2/3.3

## v0.1.205

- Add Rest/Tools: Centralize the fix for multipart/form-data PATCH requests (where parameters are not present by default)

## v0.1.204

- Remove enforced filters (as introduced in v0.1.200) and update prepared filters such that they now have a `use_policy` (is the current user
allowed to request the filter?) and a `force_use_policy` (is the usage of the filter forced for the current user?) 

## v0.1.203

- Add Rest/Query/Parameters:getBool helper function and tests
- Removed long deprecated GuzzleTools class
- Removed long deprecated Helpers\Locale class

## v0.1.202

- Rest/Query/Filter: automatically simplify the filter tree on filter creation to simplify the filter and remove
constant nodes which might not be supported by the data source system

## v0.1.201

- Rest/Query/Filter: Append a constant 'false' node when an 'inArray' condition with an empty array is appended to a filter tree

## v0.1.200

- Add enforced filters which, when configured, are always applied for collection GET requests and allows
for the restriction of results based on the logger-in user's authorization attributes

## v0.1.199

- Make QueryHelper::addFilter work for JOINS (where there are multiple entity aliases)
- Add preserveItemKeys filter to Pagination::getPage()
- Remove UserAttributeService::getUserAttribute method. Note that the service is only intended to 
deliver the currently logged-in user's attributes

## v0.1.198

- TestUserSession: allow setting up a service account user session for unit tests

## v0.1.197

- cron: print information about the cron jobs being run and their status to
  stdout instead of just logging them.
- The UserSessionInterface set up via UserAuthWebTrait and UserAuthTrait now
  behaves like a real user session in more cases, fixing various
  inconsistencies.
- Drop support for psalm

## v0.1.196

- Add method isGetRequest to AbstractStateProvider and isRootGetRequest to StateTrait
- Refactor tests

## v0.1.195

- Add Doctrine/QueryHelper utility class
- Update test utils

## v0.1.194

src/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

namespace Dbp\Relay\CoreBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('dbp_relay_core');
        $treeBuilder->getRootNode()
            ->children()
                ->scalarNode('build_info')
                    ->info('Some string identifying the current build (commit hash)')
                    ->example('deadbeef')
                ->end()
                ->scalarNode('build_info_url')
                    ->info('Some URL identifying the current build (URL to the commit on some git web interface)')
                    ->example('https://github.example.com/project/-/commit/deadbeef')
                ->end()
                ->scalarNode('logo_path')
                    ->info('Path to the logo (256x256) of the API frontend')
                    ->example('bundles/dbprelaycore/logo.png')
                ->end()
                ->scalarNode('docs_title')
                    ->info('The title text of the API docs page')
                    ->defaultValue('Relay API Gateway')
                ->end()
                ->scalarNode('docs_description')
                ->info('The description text of the API docs page (supports markdown)')
                    ->defaultValue('*part of the [digital blueprint](https://www.digital-blueprint.org) project*')
                ->end()
                ->scalarNode('queue_dsn')
                    ->info('See https://symfony.com/doc/5.3/messenger.html#redis-transport')
                    ->defaultValue('')
                    ->example('redis://redis:6379')
                ->end()
                ->scalarNode('lock_dsn')
                    ->info('https://symfony.com/doc/5.3/components/lock.html')
                    ->defaultValue('')
                    ->example('redis://redis:6379')
                ->end()
            ->end();

        return $treeBuilder;
    }
}
